File: 19/two.py
import re
import cProfile
from copy import copy


def all_indices(s, substring):
    offset = 0
    while True:
        index = s.find(substring, offset)
        if index == -1:
            break
        yield index
        offset = index + 1


inf = float('inf')
seen = set()


def contract(molecule, replacements):
    global inf, seen
    #print '>', molecule
    if molecule == 'e':
        return 0, molecule
    if 'e' in molecule or molecule in seen:
        return inf, molecule
    retval = inf, molecule
    seen.add(molecule)

    for left, cost, right in replacements:
        for start in all_indices(molecule, left):
            pre = molecule[:start]
            post = molecule[start + len(left):]
            new = pre + right + post

            #print left, cost, right, new
            rest_cost, rest_result = contract(new, replacements)
            retval = min(retval, (cost + rest_cost, rest_result), key=lambda p: p[0])
            if retval[0] < inf:
                return retval

    #print '<', molecule, retval
    print '%s\t%s\t%s' % (len(seen), len(molecule), molecule)
    return retval


def read_input(filename):
    replacements = []  # [(from, to, cost)]
    with open(filename, 'r') as f:
        lines = [line.strip() for line in f]
    input_molecule = lines.pop()
    assert lines.pop() == ''
    for line in lines:
        item = line.split(' => ')
        replacements.append((item[1], 1, item[0]))
    return input_molecule, replacements


def expand(molecule, replacements):
    molecules = set()
    for replacement in replacements:
        #print replacement
        for match in re.finditer(replacement[0], molecule):
            pre = molecule[:match.start()]
            post = molecule[match.end():]
            new = pre + replacement[2] + post
            #print molecule, pre, post, replacement, new
            molecules.add(new)
    return molecules


def expand_replacements(replacements):
    for (long, cost, short) in copy(replacements):
        for longer in expand(long, replacements):
            new = (longer, cost + 1, short)
            if new not in replacements:
                replacements.append((longer, cost + 1, short))

#print contract(*read_input('input.example3'))
#print contract(*read_input('input.example4'))
#cProfile.run("contract(*read_input('input'))")

molecule, replacements = read_input('input')
#expand_replacements(replacements)
#expand_replacements(replacements)

print contract(molecule, replacements)[0] + 1

File: 19/two.go
package main

import (
	"os"
	"math/rand"
	"fmt"
	"github.com/deckarep/golang-set"
	"io/ioutil"
	"strings"
	"runtime/pprof"
)

type Replacement struct {
	long  string
	cost  int
	short string
}

func allIndices(s string, substring string) []int {
	offset := 0
	indices := []int{}
	for {
		index := strings.Index(s[offset:], substring)
		if index == -1 {
			return indices
		}
		indices = append(indices, offset+index)
		offset += index + 1
	}
	// shuffle the indices hoping for betterer results
	for i := range indices {
		j := rand.Intn(i + 1)
		indices[i], indices[j] = indices[j], indices[i]
	}
	return indices
}

var seen = mapset.NewSet()

func contract(molecule string, replacements []Replacement) Replacement {
	//fmt.Printf("> %s\n", molecule)
	if molecule == "e" {
		return Replacement{molecule, 0, molecule}
	}
	retval := Replacement{molecule, -1, molecule}
	if strings.Contains(molecule, "e") || seen.Contains(molecule) {
		return retval
	}
	seen.Add(molecule)

	// shuffle the replacements hoping for betterer results
	for i := range replacements {
		j := rand.Intn(i + 1)
		replacements[i], replacements[j] = replacements[j], replacements[i]
	}
	for _, replacement := range replacements {
		for _, start := range allIndices(molecule, replacement.long) {
			pre := molecule[:start]
			post := molecule[start+len(replacement.long):]
			new := pre + replacement.short + post

			rest := contract(new, replacements)
			if rest.cost < len(molecule) {
				return Replacement{molecule, rest.cost + replacement.cost, rest.short}
			}
		}
	}

	fmt.Printf("%d\t%d\t%s\n", seen.Cardinality(), len(molecule), molecule)
	return retval
}

func readInput(filename string) (string, []Replacement) {
	txt, _ := ioutil.ReadFile(filename)
	lines := strings.Split(string(txt), "\n")
	inputMolecule := lines[len(lines)-1]
	replacements := []Replacement{}
	for _, line := range lines[:len(lines)-2] {
		item := strings.Split(line, " => ")
		replacements = append(replacements, Replacement{item[1], 1, item[0]})
	}
	return inputMolecule, replacements
}

func main() {
	f, _ := os.Create("cpu.prof")
	pprof.StartCPUProfile(f)
	defer pprof.StopCPUProfile()
	molecule, replacements := readInput("input")
	//fmt.Println(replacements)
	fmt.Print(contract(molecule, replacements))
}
